// botbuilder-dialogs/lib/dialogContextError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DialogContextError = void 0;
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
const dialogContext_1 = require("./dialogContext");
/**
 * An Error that includes extra dialog context, including the dialog stack.
 */
class DialogContextError extends Error {
    /**
     * Construct a DialogError.
     *
     * @param {Error | string} source Source error or error message.
     * @param {DialogContext} dialogContext Dialog context that is the source of the error.
     */
    constructor(source, dialogContext) {
        super();
        if (!(source instanceof Error) && typeof source !== 'string') {
            throw new Error('`source` argument must be an Error or a string');
        }
        if (!(dialogContext instanceof dialogContext_1.DialogContext)) {
            throw new Error('`dialogContext` argument must be of type DialogContext');
        }
        this.name = 'DialogContextError';
        if (source instanceof Error) {
            this.message = source.message;
            this.stack = source.stack;
        }
        else {
            this.message = source;
        }
        this.dialogContext = {
            activeDialog: dialogContext.activeDialog ? dialogContext.activeDialog.id : undefined,
            parent: dialogContext.parent && dialogContext.parent.activeDialog
                ? dialogContext.parent.activeDialog.id
                : undefined,
            stack: dialogContext.stack,
        };
    }
}
exports.DialogContextError = DialogContextError;
